### PR TITLE
Storage benchmark

### PR DESCRIPTION
**Pull request by hackaugusto**
## Describe your changes

This adds benchmarks for creating and reading data from the existing storages

## Checklist before requesting a review
- Repo forked and branch created from `next` according to naming convention.
- Commit messages and codestyle follow [conventions](./CONTRIBUTING.md).
- Relevant issues are linked in the PR description.
- Tests added for new functionality.
- Documentation/comments updated according to changes.

### CODE REVIEW

**Comment by hackaugusto** [comment]
Here are the results on my machine:

<details><summary>insert</summary>

for some reason the sparse merkle tree is much slower than the merkle tree. The overhead of the merkle store seems negligible in the insertion:

image

it is easier to compare when looking at a single benchmark:

```
insert/MerkleTree/new/128/128
                        time:   [600.00 µs 600.56 µs 601.18 µs]
insert/MerkleStore/with_merkle_tree/128/128
                        time:   [618.78 µs 619.33 µs 619.97 µs]
insert/SimpleSmt/new/128/128
                        time:   [38.246 ms 38.273 ms 38.302 ms]
insert/MerkleStore/with_sparse_merkle_tree/128/128
                        time:   [39.121 ms 39.151 ms 39.182 ms]
```

It seems the overhead of the Merkle Store is around 10%. And that the difference among the Merkle tree and the SMT is approximately 100x.

</details>

<details><summary>get_node</summary>

Here SMT is again super slow in comparison ... maybe there is an issue with the benchmark?

image

</details>

<details><summary>get_path</summary>

in the case of get_path the merkle store is the slowest structure:

image


</details>

</details>

File: benches/store.rs [hunk 1]
```diff
@@ -0,0 +1,207 @@
...
+            },
+        );
+    }
+}
+
+/// Benchmarks getting a node on each of the different storage backends
+///
+/// Since MerkleTree can only be created when a power-of-two number of elements is used, the sample
+/// sizes are limited to that
+fn get_node(c: &mut Criterion) {
```
Comment (bobbinth) [inline benches/store.rs:98]: I would probably break this into two separate comparisons:

1. `MerkleTree` vs. `MerkleStore` instantiated with `with_merkle_tree`.
2. `SimpleSmt` of depth 63 vs. `MerkleStore` instantiated with `with_sparse_merkle_tree` (I believe this will assume depth 63 as well, but I wonder if we should make the depth be an explicit parameter).

File: benches/store.rs [hunk 2]
```diff
@@ -0,0 +1,207 @@
...
+            )
+        });
+    }
+}
+
+/// Benchmarks getting a path on each of the different storage backends
+///
+/// Since MerkleTree can only be created when a power-of-two number of elements is used, the sample
+/// sizes are limited to that
+fn get_path(c: &mut Criterion) {
```
Comment (bobbinth) [inline benches/store.rs:154]: Similar comment here - I'd have two separate comparisons:

1. `MerkleTree` vs. `MerkleStore` instantiated with `with_merkle_tree`.
2. `SimpleSmt` of depth 63 vs. `MerkleStore` instantiated with `with_sparse_merkle_tree`.

**Review by bobbinth — CHANGES_REQUESTED** [review]
Thank you! Looks good! I left a few comments inline.

Regarding results:
* **insert** - I think the result makes sense. Regular Merkle tree is very small (depth 7 for 128 leaves) but SMTs are big (depth 63) - so, there is much more hashing involved in SMTs.
* **get_node** - I think the results here make sense as well. For simple Merkle tree this is a single memory look up - not traversals needed. So, it is super fast. For Merkle store, it is also just lookups - but now we need to do more of them and also comparisons are a bit complicated (not sure if this justifies the full delta though). For Simple SMT, we are hitting a branch node, and the way it is currently implemented, this requires one invocation of RPO, which takes about 5us. If we were to hit a leaf node, I think this would work much faster.
* **get_path** - here I'm a bit surprised that delta between Merkle store and simple SMT is so high. I guess the result for Merkle store is consistent with **get_node** result, but it is still interesting that traversal in simple SMT can be done so much faster. I wonder what drives it - e.g., bit iterator, putting more complex objets into a `BTreeMap`, something else?

Lastly, would be great to see similar benchmarks for `set_node`.

File: benches/store.rs [hunk 3]
```diff
@@ -0,0 +1,207 @@
...
+use rand_utils::{rand_array, rand_value};
+
+static BATCH_SIZES: [usize; 7] = [
+    2usize.pow(1),
+    2usize.pow(2),
+    2usize.pow(3),
+    2usize.pow(4),
+    2usize.pow(5),
+    2usize.pow(6),
+    2usize.pow(7),
```
[the remark this answers is not in the document]
Comment (hackaugusto) [reply benches/store.rs:14]: changed

File: benches/store.rs [hunk 4]
```diff
@@ -0,0 +1,207 @@
...
+    let random_data_size = BATCH_SIZES.into_iter().max().unwrap();
+    let random_data: Vec<RpoDigest> = (0..random_data_size)
+        .map(|_| rand_array::<Felt, 4>().into())
+        .collect();
+
+    for size in BATCH_SIZES {
+        let leaves = &random_data[..size];
+
+        // MerkleTree constructor is optimized to work with vectors. Create a new copy of the data
+        // nd pass it to the benchmark function
```
[the remark this answers is not in the document]
Comment (hackaugusto) [reply benches/store.rs:33]: thanks, fixed :)

File: benches/store.rs (hunk 1 above)
Comment (hackaugusto) [reply benches/store.rs:98]: split benchmarks

File: benches/store.rs (hunk 2 above)
Comment (hackaugusto) [reply benches/store.rs:154]: split benchmarks

**Comment by bobbinth** [comment]
Here is the results I'm getting for size 1024 on my machine:

#### get_node
This should really be called **get_leaf** because we are testing getting leafs from the data structure rather than internal nodes.

|                                       |  Specialized | MerkleStore |
| ---------------------- | :----------: | ------------ |
| MerkleTree (depth 10) | 20 ns          | 1 us               |
| SMT (depth 63)            | 52 ns         |  5.8 us           |

I think these numbers make sense. For specialized implementations, this is just one lookup either in memory or in a `BTreeMap` while for `MerkleStore` it requires tree traversal. Though I do wonder if 1 us maybe a bit too high for doing 10 `BTreeMap` lookups and comparisons.

#### get_path

|                                       |  Specialized | MerkleStore |
| ---------------------- | :----------: | ------------ |
| MerkleTree (depth 10) | 68 ns          | 1.2 us               |
| SMT (depth 63)            | 700 ns        |  680 ns           |

These numbers look odd to me. Specifically, it is strange that building a Merkle path is 10x faster than getting a node.

File: benches/store.rs [hunk 5]
```diff
@@ -0,0 +1,257 @@
...
+            .unwrap();
+        let store = MerkleStore::new()
+            .with_sparse_merkle_tree(smt_leaves)
+            .unwrap();
+        let depth = smt.depth();
+        let root = smt.root();
+
+        group.bench_function(BenchmarkId::new(format!("SimpleSmt"), size), |b| {
+            b.iter_batched(
+                || rand_value::<u64>() % u64::try_from(size).unwrap(),
```
Comment (bobbinth) [inline benches/store.rs:150]: I wonder if here, as well as in the equivalent `MerkleStore` benchmark below, we should make it so that sometimes we hit empty nodes (i.e., maybe mod by `size * 2` or something like that?).

File: benches/store.rs [hunk 6]
```diff
@@ -0,0 +1,257 @@
...
+            )
+        });
+
+        group.bench_function(BenchmarkId::new(format!("MerkleStore"), size), |b| {
+            b.iter_batched(
+                || rand_value::<u64>() % u64::try_from(size).unwrap(),
+                |value| black_box(store.get_path(root, NodeIndex::new(depth, value))),
+                BatchSize::SmallInput,
+            )
+        });
```
Comment (bobbinth) [inline benches/store.rs:245]: As mentioned in https://github.com/0xPolygonMiden/crypto/pull/97#issuecomment-1474698263, assuming I understand it correctly, this benchmark seems to produce very counter-intuitive results. We should try to understand why.

Also, we should probably test building paths leading to empty nodes as well (similar to my other comment above).

Lastly, the ID of the benchmark conflicts with another benchmark above - maybe change it slightly here?

**Review by bobbinth — DISMISSED** [review]
Thank you! Looks good! I left a few comments inline, but beyond these:

* I'd probably rename current `get_node_*` into `get_leaf_*`.
* I'd probably add `get_node_*` benchmarks which would retrieve a node which is not a leaf (i.e., a node at depth/2 for a tree).
* I think we should also have a benchmark for `set_leaf`.

File: benches/store.rs [hunk 7]
```diff
@@ -0,0 +1,257 @@
...
+    for size in BATCH_SIZES {
+        let leaves = &random_data[..size];
+
+        let mtree_leaves: Vec<Word> = leaves.iter().map(|v| v.into()).collect();
+        let mtree = MerkleTree::new(mtree_leaves.clone()).unwrap();
+        let store = MerkleStore::new().with_merkle_tree(mtree_leaves).unwrap();
+        let depth = mtree.depth();
+        let root = mtree.root();
+
+        group.bench_function(BenchmarkId::new(format!("MerkleTree"), size), |b| {
```
[the remark this answers is not in the document]
Comment (hackaugusto) [reply benches/store.rs:101]: ah, thanks! I remove the size from the string and didn't see the format was no longer necessary, changed.

File: benches/store.rs [hunk 8]
```diff
@@ -0,0 +1,257 @@
...
+            },
+        );
+    }
+}
+
+/// Benchmarks getting a node on each of the different storage backends
+///
+/// Since MerkleTree can only be created when a power-of-two number of elements is used, the sample
+/// sizes are limited to that
+fn get_node_merkletree(c: &mut Criterion) {
```
[the remark this answers is not in the document]
Comment (hackaugusto) [reply benches/store.rs:84]: done

File: benches/store.rs (hunk 5 above)
Comment (hackaugusto) [reply benches/store.rs:150]: maybe a new bench test only for empty nodes? that way we can tell the performance of each case individually

File: benches/store.rs (hunk 6 above)
Comment (hackaugusto) [reply benches/store.rs]: changed the benchmark group names to `get_path_merkletree` and `get_path_simplesmt`

File: benches/store.rs [hunk 9]
```diff
@@ -0,0 +1,353 @@
...
+            )
+        });
+    }
+}
+
+/// Benchmarks getting an invalid index on the MerkleTree and MerkleStore backends.
+///
+/// Since MerkleTree can only be created when a power-of-two number of elements is used, the sample
+/// sizes are limited to that
+fn get_leaf_merkletree_invalid(c: &mut Criterion) {
```
Comment (bobbinth) [inline benches/store.rs:123]: I'm not sure we need this benchmark. I mean it doesn't hurt to have it, but also, this is probably a rather edge case.

File: benches/store.rs [hunk 10]
```diff
@@ -0,0 +1,353 @@
...
+                || rand_value::<u64>() % u64::try_from(size).unwrap(),
+                |value| black_box(store.get_node(root, NodeIndex::new(depth, value))),
+                BatchSize::SmallInput,
+            )
+        });
+    }
+}
+
+/// Benchmarks getting an invalid index on the SMT and MerkleStore backends.
+fn get_leaf_simplesmt_invalid(c: &mut Criterion) {
```
Comment (bobbinth) [inline benches/store.rs:210]: I think what we want to benchmark here is not invalid access, but accessing an empty node. Since we are building an SMT with relatively few elements (i.e., at most 1024) almost any random index should yield an empty node - so, we probably don't need to iterate over batch size here.

**Review by bobbinth — CHANGES_REQUESTED** [review]
Thank you, Looks good! I left a couple of inline comments. Also, any reason you are not adding benchmarks for `set_node()`?

Overall, my takeaway is that using `MerkleStore` is about 10x slower than `SimpleSmt` and almost 100x slower than using `MerkleTree`.

This probably doesn't matter for transactions too much, but might matter for batch and block construction. So, in the future, we might want to build specialized advice providers which work with `MerkleTree`'s, `SimpleSmt`'s and probably `Mmr`'s whenever possible.

File: benches/store.rs (hunk 9 above)
Comment (hackaugusto) [reply benches/store.rs:123]: this was requested here https://github.com/0xPolygonMiden/crypto/pull/97#discussion_r1141613289 , should I remove it?

Comment (hackaugusto) [reply benches/store.rs]: removed. I had understood that the benchmark was for invalid indexes, not for empty hashes, benchmark is updated

File: benches/store.rs (hunk 10 above)
Comment (hackaugusto) [reply benches/store.rs]: added

**Review by bobbinth — APPROVED** [review]
All looks good! Thank you!
